Document and implement Murmur locking strategy

mkrautz opened the pull request:
[I am ready to have a few more eyes on this PR now... Don't mind the refactoring commits -- I can split them out later, but for now, the later locking changes depend on them. Also don't mind the CMakeLists.txt -- it is only there to help you if you want to review using CLion... It will be removed before landing. It's not in a state where it can be landed at the moment.]

This PR documents Murmur's locking strategy in docs/MurmurLocking.md.

It also implements the locking scheme described in docs/MurmurLocking.md in places where Murmur was not living up to the requirements.

The best way to approach this for review is to read docs/MurmurLocking.md first.

Then, look through each commit that adds locking to a part of Murmur, and ensure it is correct according to the MurmurLocking.md document.

Let us, as an example, use the commit that adds a write lock to addChannel/removeChannel/link/unlink in the main thread. To review this commit, you would ensure that all code paths in the main thread that write to the data used by those methods take a write lock on qrwlVoiceThread. You could, for example, use "Find usages" in CLion, or equivalents from other IDEs.

To properly review the whole thing, you would need to:

- Read the docs/MurmurLocking.md document, and note the methods that the voice thread consists of.
- Go through all methods uesd in Murmur's voice thread, and take notes about what kind of data the methods access that is potentially a data race. (My notes are at https://gist.github.com/mkrautz/952fddc88e764a4b7179efd4be95f7b6 -- this is what became MurmurLocking.md -- but it's more helpful if you make your own, in case I left something out.)
- Check that all the data accesses that you found are properly documented in docs/MurmurLocking.md. If not, raise it as an issue in the PR. That document is meant to cover *everything* that can be touched by both the voice thread and and the main thread at the same time.
- Then, go through all data mentioned in docs/MurmurLocking.md and ensure that the locking for that data is correct.

mkrautz commented:
docs/MurmurLocking.md is missing some crucial details about how the voice thread uses qrwlVoiceThread: it holds a read lock to it during most of its operation.

mkrautz commented:
@hacst @Kissaki @fwaggle 

Anyone up for reviewing this?

We'll probably need to do multiple iterations...

mkrautz commented:
> docs/MurmurLocking.md is missing some crucial details about how the voice thread uses qrwlVoiceThread: it holds a read lock to it during most of its operation.

Fixed.

hacst commented on src/murmur/Messages.cpp:
```diff
@@ -95,43 +95,45 @@ void Server::msgAuthenticate(ServerUser *uSource, MumbleProto::Authenticate &msg
-		        (u->qsName.toLower() == uSource->qsName.toLower())) {
-			uOld = u;
-			break;
+			foreach(ServerUser *u, qhUsers) {
```
Seems like indendation got a bit messed up here.

mkrautz replied:
Fixed.

mkrautz commented:
https://github.com/mumble-voip/mumble/pull/2276/commits/efbd24d2525d88a26695624b37ca62e156128e0d has a lot of unintended whitespace changes... I will fix. (I blame CLion.)

mkrautz commented:
Fixed the aforementioned whitespace issues.

PTAL.

mkrautz commented:
Anything in particular I can do to help to make this easier to review? :(

hacst commented:
Nag me more? Honestly. It is know to work on me ;)

mkrautz commented:
OK! :-)

@hacst Ping!

hacst commented:
It's starting to work. Keep at it!!11 ;) Let's set me a deadline for this weekend. I do realize that made me delay you landing this for ages but I won't be able to get a big enough block of free time to really tackle such mindbeding stuff till then. Hope that's ok.

mkrautz commented:
Sure thing!

mkrautz commented:
@hacst Ping?

hacst commented:
Hi. Sorry for the wait. The only things that stood out to me locking/threading wise were:

* I think bRunning in Server.h ought to be an atomic as it is read from the voice thread (::run) but set from the main thread. In theory (though unlikely in practice) I think this could confuse the optimizer.
* In Server.cpp there is one case where the voice thread takes the qrwlVoiceThread for writing using ```qrwlVoiceThread.lockForWrite()```. This confused me a bit since from the documentation I assumed only the main thread is allowed to do that. The use itself looked safe to me though.

Apart from those I couldn't find anything violating the rules you set/derived (which sound good to me).

Note that I didn't look too closely at the GRPC related code. I haven't built murmur with that yet and it confused CLion when I tried to enable it. I assume this isn't that big of a deal since it's somewhat separate and not built in by default at this time. However it does create its own thread so it might warrant a closer look before we ship it as a "stable" thing.

TL;DR: LGTM apart from the two bullet points above ;)

mkrautz commented:
Regarding bRunning: Yep, that's true. But that's also true for Mumble. We use it a lot in Mumble.

Qt 5 has its own atomic implementation in QThread (isInterruptionRequested, requestInterruption), but I really think we should do something about those as well...

Thanks for the review.

mkrautz commented:
@hacst isn't the case where qrwlVoiceThread is locked for writing in the voice thread covered by https://github.com/mumble-voip/mumble/pull/2276/commits/5f5f85fca693853882f56330f3c1bf23d5f84675#diff-a39263b917c72193a89943cd67f5e226R165?

hacst commented:
Hm. The location is unrelated to qmTargetCache though? Isn't it. I'm talking about ```foreach(ServerUser *usr, qhHostUsers.value(ha))```.

Re-reading my comment I forgot an important part about that use of ```qrwlVoiceThread.lockForWrite()```. I couldn't see where it was released in case a race had occured and the user was no longer present.

mkrautz commented:
> Hm. The location is unrelated to qmTargetCache though? Isn't it. I'm talking about ```foreach(ServerUser *usr, qhHostUsers.value(ha))```.

Thanks!

This section:
https://github.com/mumble-voip/mumble/pull/2276/commits/5f5f85fca693853882f56330f3c1bf23d5f84675#diff-a39263b917c72193a89943cd67f5e226R129

is missing qhHostUsers and qhPeerUsers.

mkrautz commented:
> Re-reading my comment I forgot an important part about that use of qrwlVoiceThread.lockForWrite(). I couldn't see where it was released in case a race had occured and the user was no longer present.

Do you mean this code?

https://github.com/mumble-voip/mumble/blob/master/src/murmur/Server.cpp#L1106-L1113

mkrautz commented:
>> Re-reading my comment I forgot an important part about that use of qrwlVoiceThread.lockForWrite(). I couldn't see where it was released in case a race had occured and the user was no longer present.
>
> Do you mean this code?
> 
> https://github.com/mumble-voip/mumble/blob/master/src/murmur/Server.cpp#L1106-L1113

I guess you mean this:

https://github.com/mumble-voip/mumble/blob/master/src/murmur/Server.cpp#L828

mkrautz commented:
^- that case does look suspicious... I'll have a look tomorrow.

mkrautz commented:
@hacst PTAL

I have not fixed the bRunning issue, as it also has ties to Mumble. I think we should fix that separately.

I have added qhPeerUsers and qhHostUsers to MurmurLocking.md (amended to https://github.com/mumble-voip/mumble/pull/2276/commits/b289c17cd26d9fb0bd218f971f6ba4c1699a4e43)
I have added a commit to fix the issue you raised regarding locking in the unknown peer path in Server::run: https://github.com/mumble-voip/mumble/pull/2276/commits/3dc8b567219fc1a59aeae545221ac962d51005fd

hacst commented:
The classification for ```qhPeerUsers``` and ```qhHostUsers``` still looks wrong to me.  ```qhPeerUsers``` is shared data as it is inserted into from ```Server::run``` using ```qhPeerUsers.insert(key, u);``` and ```Server::connectionClosed``` removes from it from the main thread. For ```qhHostUsers``` while the container itself is (hopefully, hashmap auto-creation semantics are dangerous) not modified by the voice thread it does modify the contained data which assume ought to share the ownership? In that case it would also be shared data.

The unlock seems safe now.

mkrautz commented:
Agreed. I put them in the wrong section...

mkrautz commented:
Moved them to shared section.

PTAL @hacst.

hacst commented:
LGTM.

mkrautz commented:
Thanks for the review!

I removed the CMake commit and merged it.
